rust/hyperlane-base/src/settings/chains.rs

```diff
@@ -57,81 +79,236 @@ pub enum TransactionSubmissionType {
-            .and_then(|s| s.try_into().ok())
-            .unwrap_or(1999)
+/// Indexing settings
+#[derive(Debug, Default, Clone)]
+pub struct IndexSettings {
+    /// The height at which to start indexing contracts.
+    pub from: u32,
+    /// The number of blocks to query at once when indexing contracts.
+    pub chunk_size: u32,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct RawIndexSettings {
+    from: Option<StrOrInt>,
+    chunk: Option<StrOrInt>,
+}
+
+impl FromRawConf<'_, RawIndexSettings> for IndexSettings {
+    fn from_config_filtered(
+        raw: RawIndexSettings,
+        cwp: &ConfigPath,
+        _filter: (),
+    ) -> ConfigResult<Self> {
+        let mut err = ConfigParsingError::default();
+
+        let from = raw
+            .from
+            .and_then(|v| v.try_into().take_err(&mut err, || cwp + "from"))
+            .unwrap_or_default();
+
+        let chunk_size = raw
+            .chunk
+            .and_then(|v| v.try_into().take_err(&mut err, || cwp + "chunk"))
+            .unwrap_or(1999);
+
+        err.into_result()?;
+        Ok(Self { from, chunk_size })
     }
 }
 
 /// A chain setup is a domain ID, an address on that chain (where the mailbox is
 /// deployed) and details for connecting to the chain API.
-#[derive(Clone, Debug, Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct ChainSetup {
-    /// Chain name
-    pub name: String,
-    /// Chain domain identifier
-    pub domain: StrOrInt,
+#[derive(Clone, Debug)]
+pub struct ChainConf {
+    /// The domain
+    pub domain: HyperlaneDomain,
     /// Signer configuration for this chain
     pub signer: Option<SignerConf>,
     /// Number of blocks until finality
-    pub finality_blocks: StrOrInt,
+    pub finality_blocks: u32,
     /// Addresses of contracts on the chain
     pub addresses: CoreContractAddresses,
     /// The chain connection details
-    #[serde(flatten, default)]
     pub connection: Option<ChainConnectionConf>,
     /// How transactions to this chain are submitted.
-    #[serde(default)]
     pub txsubmission: TransactionSubmissionType,
     /// Configure chain-specific metrics information. This will automatically
     /// add all contract addresses but will not override any set explicitly.
     /// Use `metrics_conf()` to get the metrics.
-    #[serde(default)]
     pub metrics_conf: PrometheusMiddlewareConf,
     /// Settings for event indexing
-    #[serde(default)]
     pub index: IndexSettings,
 }
 
-impl ChainSetup {
+/// A raw chain setup is a domain ID, an address on that chain (where the
+/// mailbox is deployed) and details for connecting to the chain API.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct RawChainConf {
+    name: Option<String>,
+    domain: Option<StrOrInt>,
+    pub(super) signer: Option<RawSignerConf>,
+    finality_blocks: Option<StrOrInt>,
+    addresses: Option<RawCoreContractAddresses>,
+    #[serde(flatten, default)]
```

Comment:
Yes, the content of connection is actually in a separate structure using `serde(content = "connection")` This was done so the tag `protocol` can be in the root.

Whether we like this choice or not, making a change would break the current config format.